### PR TITLE
Merge ECC and ECC_MOCK

### PR DESCRIPTION
**Pull request by mbrandenburger**
- Implements the new flows for init/invoke and endorse
- remove old ecc
- tested with test-network and test go client (currently endorse does not succeed as `ecc_enclave` does not set `enclaveId` in response yet)

Signed-off-by: bur <bur@zurich.ibm.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing changes and/or breaks backward compatability?**:

```

### CODE REVIEW

File: examples/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Copyright IBM Corp. All Rights Reserved.
```
Comment (g2flyer) [inline examples/Dockerfile:1]: this should be obsolete?

File: examples/auction/Makefile [hunk 2]
```diff
@@ -20,3 +20,7 @@ build: $(BUILD_DIR)
 
 clean:
 	rm -rf $(BUILD_DIR)
+
+docker:
+	$(DOCKER) build -f ../Dockerfile -t fpc/fpc-auction .
+
```
Comment (g2flyer) [inline examples/auction/Makefile:26]: This also should be obsolete?

File: examples/echo/Makefile [hunk 3]
```diff
@@ -20,3 +20,6 @@ build: $(BUILD_DIR)
 
 clean:
 	rm -rf $(BUILD_DIR)
+
+docker:
+	$(DOCKER) build -f ../Dockerfile -t fpc/fpc-echo .
```
Comment (g2flyer) [inline examples/echo/Makefile:25]: obsolete?

**Review by g2flyer — COMMENTED** [review]
Find various comments in source.  Overall looks good. I didn't manage to test end-to-end. However, i did figure out your docker problem. $FPC_PATH/.dockerignore is the culprit.  I've updated makefile, dockerfile and readme to make that better and also implemented the gotags for enclave vs mock_enclave.  I've added the format-patch below but to be on the save side (i.e., no copy/paste) i also sent it to you as a separate email .

One other comment i couldn't put inline:
- why do we not keep `ecc/chaincode/crypto/ecdsa.go` in `internal/utils/` where we already have crypto stuff? Right now we have closely related crypto go libraries at three different places (here, utils and client_sdk) which is rather counter-intuitive ...

```
From 24b94a89896a1a0bee5451d77b84997b01891078 Mon Sep 17 00:00:00 2001
From: Michael Steiner <michael.steiner@intel.com>
Date: Fri, 11 Dec 2020 03:51:31 +0000
Subject: [PATCH] Fixed & improved fpc-app docker & implement gotags for
 mock_enclave

Signed-off-by: Michael Steiner <michael.steiner@intel.com>
---
 ecc/Dockerfile.fpc-app                | 11 ++++-------
 ecc/Makefile                          | 13 ++++++-------
 ecc/chaincode/enclave/enclave.go      | 10 +++++++++-
 ecc/chaincode/enclave/logging_cgo.go  |  2 ++
 ecc/chaincode/enclave/logging_go.go   |  2 ++
 ecc/chaincode/enclave/mock_enclave.go |  7 +++++++
 ecc/chaincode/enclave/shim.go         |  9 +++++++++
 integration/test-network/README.md    |  8 ++------
 8 files changed, 41 insertions(+), 21 deletions(-)

diff --git a/ecc/Dockerfile.fpc-app b/ecc/Dockerfile.fpc-app
index f6ce76c..13fec10 100644
--- a/ecc/Dockerfile.fpc-app
+++ b/ecc/Dockerfile.fpc-app
@@ -8,21 +8,18 @@
 #
 #  Configuration (build) paramaters (for defaults, see below section with ARGs)
 #  - fpc image version:                                         FPC_VERSION
-#  - path to directory containing chaincode's enclave.so:       enclave_so_path
+#
+# expects to be run with the docker context pointed o directory containing chaincode's enclave.so

 ARG BOILERPLATE_EXTENSION
 ARG FPC_VERSION=latest

 FROM hyperledger/fabric-private-chaincode-boilerplate-ecc:${FPC_VERSION}

-ARG enclave_so_path
-

 ARG CC_PATH="/usr/local/bin"
 ARG CC_LIB_PATH=${CC_PATH}"/enclave/lib"

-RUN test -n "$enclave_so_path"
-
-copy ${enclave_so_path}/enclave.signed.so ${CC_LIB_PATH}/
+COPY /enclave.signed.so ${CC_LIB_PATH}/

-copy ${enclave_so_path}/mrenclave ${CC_PATH}/
+COPY /mrenclave ${CC_PATH}/
diff --git a/ecc/Makefile b/ecc/Makefile
index 684e69a..5e3a024 100644
--- a/ecc/Makefile
+++ b/ecc/Makefile
@@ -63,14 +63,13 @@ docker-boilerplate-ecc: ecc
        $(DOCKER) tag $(DOCKER_BOILERPLATE_ECC_IMAGE) $(DOCKER_BOILERPLATE_ECC_IMAGE):${FPC_VERSION}

 docker-fpc-app: docker-boilerplate-ecc
-#      if [ -z "$(DOCKER_IMAGE)" ]; then\
-#              echo "\033[0;31mERROR: cannot override $(CC_NAME) docker image - not found\033[0m";\
-#              exit 1;\
-#      fi
+       if [ -z "$(DOCKER_IMAGE)" ]; then\
+               echo "\033[0;31mERROR: cannot override $(CC_NAME) docker image - not found\033[0m";\
+               exit 1;\
+       fi
        echo "\033[1;33mWARNING: overriding $(DOCKER_IMAGE) docker image\033[0m"
        $(DOCKER) build $(DOCKER_BUILD_OPTS) -t $(DOCKER_IMAGE) -f Dockerfile.fpc-app \
-               --build-arg enclave_so_path=$(DOCKER_ENCLAVE_SO_PATH) \
-               ..
+               $(DOCKER_ENCLAVE_SO_PATH)

 docker-run:
        if [ "$(SGX_MODE)" = "HW" ]; then \
@@ -106,4 +105,4 @@ docker-clean: docker-stop
 clobber: docker-clean
 clean: docker-clean
        $(GO) clean
-       $(RM) ecc enclave/ecc-enclave-lib enclave/ecc-enclave-include
\ No newline at end of file
+       $(RM) ecc enclave/ecc-enclave-lib enclave/ecc-enclave-include
diff --git a/ecc/chaincode/enclave/enclave.go b/ecc/chaincode/enclave/enclave.go
index cc39027..423663f 100644
--- a/ecc/chaincode/enclave/enclave.go
+++ b/ecc/chaincode/enclave/enclave.go
@@ -1,3 +1,12 @@
+// +build !mock_ecc
+
+/*
+Copyright IBM Corp. All Rights Reserved.
+Copyright 2020 Intel Corporation
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
 package enclave

 import "C"
@@ -31,7 +40,6 @@ type EnclaveStub struct {
        sem *semaphore.Weighted
 }

-// TODO enabled GOTAGS-based toggle to switch between this EnclaveStub and MockStub
 // NewEnclave starts a new enclave
 func NewEnclaveStub() StubInterface {
        return &EnclaveStub{sem: semaphore.NewWeighted(8)}
diff --git a/ecc/chaincode/enclave/logging_cgo.go b/ecc/chaincode/enclave/logging_cgo.go
index 1dc303c..324b3ea 100644
--- a/ecc/chaincode/enclave/logging_cgo.go
+++ b/ecc/chaincode/enclave/logging_cgo.go
@@ -1,3 +1,5 @@
+// +build !mock_ecc
+
 /*
    Copyright 2019 Intel Corporation
    Copyright IBM Corp. All Rights Reserved.
diff --git a/ecc/chaincode/enclave/logging_go.go b/ecc/chaincode/enclave/logging_go.go
index 32c2955..6845dea 100644
--- a/ecc/chaincode/enclave/logging_go.go
+++ b/ecc/chaincode/enclave/logging_go.go
@@ -1,3 +1,5 @@
+// +build !mock_ecc
+
 /*
    Copyright 2019 Intel Corporation
    Copyright IBM Corp. All Rights Reserved.
diff --git a/ecc/chaincode/enclave/mock_enclave.go b/ecc/chaincode/enclave/mock_enclave.go
index 457873b..bdcdb37 100644
--- a/ecc/chaincode/enclave/mock_enclave.go
+++ b/ecc/chaincode/enclave/mock_enclave.go
@@ -1,3 +1,5 @@
+// +build mock_ecc
+
 /*
 Copyright IBM Corp. All Rights Reserved.
 Copyright 2020 Intel Corporation
@@ -29,6 +31,11 @@ type MockEnclaveStub struct {
        enclaveId  string
 }

+// NewEnclave starts a new enclave
+func NewEnclaveStub() StubInterface {
+       return &MockEnclaveStub{}
+}
+
 func (m *MockEnclaveStub) Init(serializedChaincodeParams, serializedHostParamsBytes, serializedAttestationParams []byte) ([]byte, error) {

        hostParams := &protos.HostParameters{}
diff --git a/ecc/chaincode/enclave/shim.go b/ecc/chaincode/enclave/shim.go
index 256f501..897651b 100644
--- a/ecc/chaincode/enclave/shim.go
+++ b/ecc/chaincode/enclave/shim.go
@@ -1,3 +1,12 @@
+// +build !mock_ecc
+
+/*
+Copyright IBM Corp. All Rights Reserved.
+Copyright 2020 Intel Corporation
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
 package enclave

 import (
diff --git a/integration/test-network/README.md b/integration/test-network/README.md
index 50b5528..8f9877f 100644
--- a/integration/test-network/README.md
+++ b/integration/test-network/README.md
@@ -9,13 +9,9 @@ Make sure you have installed [yq](https://github.com/mikefarah/yq).
 Before you start the network make sure you build ercc and ecc. In

 ```bash
-cd $FPC_PATH/ercc
-make && make docker
+make -C $FPC_PATH/ercc all docker

-cd $FPC_PATH/ecc
-export DOCKER_IMAGE=fpc/fpc-echo
-export DOCKER_ENCLAVE_SO_PATH=examples/echo/_build/lib
-make && make docker
+make -C $FPC_PATH/ecc DOCKER_IMAGE=fpc/fpc-echo DOCKER_ENCLAVE_SO_PATH=$FPC_PATH/examples/echo/_build/lib all docker
 ```

 ## Prepare network
--
2.17.1
```

File: ecc/chaincode/enclave/interface.go [hunk 4]
```diff
@@ -10,12 +10,15 @@ package enclave
...
+	"github.com/hyperledger/fabric/common/flogging"
 )
 
+var logger = flogging.MustGetLogger("enclave")
+
 type StubInterface interface {
 
 	// triggered by an admin
-	Init(chaincodeParams *protos.CCParameters, hostParams *protos.HostParameters, attestationParams []byte) ([]byte, error)
+	Init(chaincodeParams, hostParams, attestationParams []byte) ([]byte, error)
```
[the remark this answers is not in the document]
Comment (mbrandenburger) [reply ecc/chaincode/enclave/interface.go:21]: good point, let me check

File: ecc/chaincode/enclave/enclave.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+	// TODO give this signed proposal to enclave
+	serializedProposal, err := proto.Marshal(proposal)
+	if err != nil {
+		return nil, err
+	}
+	_ = serializedProposal
+
+	// args
+	jsonArgs, err := json.Marshal(stub.GetStringArgs())
```
[the remark this answers is not in the document]
Comment (mbrandenburger) [reply ecc/chaincode/enclave/enclave.go]: yes

File: ecc/chaincode/enclave/interface.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (g2flyer) [reply ecc/chaincode/enclave/interface.go:21]: and what was the outcome of the check ;-?

**Review by g2flyer — APPROVED** [review]
Addresses my prior comments and sample-net echo tests work in SIM mode for both enclave (with expected endorse failure) and mock_enclave (ie.. fpc-echo docker images built as `make -C $FPC_PATH/ecc DOCKER_IMAGE=fpc/fpc-echo DOCKER_ENCLAVE_SO_PATH=$FPC_PATH/examples/echo/_build/lib clean all docker GOTAGS="-tags mock_ecc"`).  It doesn't work with HW mode. I do have corresponding patches which i will add as a later PR ...

**Review by bvavala — APPROVED** [review]
works/fails as expected.
However, I am unable to test with the test network.
Merging for now, and solve this later.

**Pull request merged**